jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DefaultManyToManyShortestPaths.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * (C) Copyright 2019-2019, by Semen Chudakov and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.shortestpath;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphPath;
+import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Naive algorithm for many-to-many shortest paths problem.
+ *
+ * <p>
+ * Complexity of the algorithm is $O(|S|)(V\log V + E)$, where $S$ is the set of source vertices,
+ * $V$ is the set of graph vertices and $E$ is the set of graph edges of the graph.
+ *
+ * <p>
+ * For each source vertex a single source shortest paths search is performed, which is stopped
+ * as soon as all target vertices are reached. Shortest paths trees is constructed using
+ * {@link DijkstraClosestFirstIterator}.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author Semen Chudakov
+ */
+public class DefaultManyToManyShortestPaths<V, E> extends BaseManyToManyShortestPaths<V, E> {
```

Comment:
This class makes sense only if you also provide the user with a choice of algorithm.

I would suggest to add a constructor with a `Function<Graph<V,E>, ShortestPathAlgorithm<V,E>>` parameter, which would be the supplier of the shortest path algorithm.  Moreover, also provide a constructor with a default choice which should be bidirectional Dijkstra.

Reply:
I would like to make some clarifications on this issue. 

The idea behind the `DefaultManyToManyShortestPaths` is to provide a way to compute many-to-many shortest paths without computing Contraction Hierarchy (which is time consuming for dense graphs). 

To my knowledge, there are 3 ways to solve many-to-many shortest paths problem without CH. 
(`m = size(sources)`, `n = size(targets)`)

1. Perform `O(m*n)` queries with `ShortestPathAlgorithm#getPath(source, target)` method.
2. Perform `O(m)` queries with `ShortestPathAlgorithm#getPaths(source)` method.   
3. `O(m)` times use `DijkstraClosestFirstIterator` to iterate from every source vertex to all targets and no further. 

I have chosen the last approach for performance reasons and implemented it in `BaseManyToManyShortestPaths#getShortestPathsTree` which is also used in `CHManyToManyShortestPath` class. Unlike the first two approaches the third one does not allow to parameterize the shortest path algorithm implementation as we do not have a dedicated interface for `-ClosestFirstIterator` implementations. 

Would you like me to implement the `DefaultManyToManyShortestPaths` with one of the first to approaches?

Reply:
You are right, thanks for the clarification.

Some ideas if you have the time and energy:
 - It might make sense to adjust the 3rd way to use `min(m, n)` times the iterator. It might be faster for some cases. 
 - No need to implement the 2nd way. The 3rd should be faster.
 - The 1st way would be nice to have (perhaps in a separate class called `BidirectionalManyToManyShortestPaths`) using m*n queries of bidirectional Dijkstra.